webapp/src/serial.tsx

```diff
@@ -0,0 +1,309 @@
+/// <reference path="../../localtypings/smoothie.d.ts" />
+
+import * as React from "react"
+import * as pkg from "./package"
+import * as core from "./core"
+import * as srceditor from "./srceditor"
+import * as sui from "./sui"
+import * as codecard from "./codecard"
+import * as canvaschart from "./canvaschart"
+
+import Cloud = pxt.Cloud
+import Util = pxt.Util
+
+const lf = Util.lf
+
+export class Editor extends srceditor.Editor {
+    charts: Chart[] = []
+    chartIdx: number = 0
+    consoleBuffer: string = ""
+    // TODO pass these values in with props or config
+    isSim: boolean = true
+    maxConsoleLineLength: number = 500
+    maxConsoleEntries: number = 100
+    active: boolean = true
+    rawDataBuffer: string = ""
+    //TODO reasonable buffer size?
+    maxBufferLength: number = 5000
+
+    getId() {
+        return "serialEditor"
+    }
+
+    acceptsFile(file: pkg.File) {
+        return file.name === pxt.SERIAL_EDITOR_FILE
+    }
+
+    isGraphable(v: string) {
+        return /[a-z]*:[0-9.]*/.test(v)
+    }
+
+    setSim(b: boolean) {
+        this.isSim = b
+        this.clear()
+    }
+
+    constructor(public parent: pxt.editor.IProjectView) {
+        super(parent)
+        window.addEventListener("message", this.processMessage.bind(this), false)
+    }
+
+    processMessage(ev: MessageEvent) {
+        let msg = ev.data
+        if (this.active && msg.type === "serial") {
+            //TODO y tho
+            const smsg = msg as pxsim.SimulatorSerialMessage
+            const sim = !!smsg.sim
+            if (sim == this.isSim) {
+                const data = smsg.data || ""
+                const source = smsg.id || "?"
+                //TODO not using theme anymore
```

Comment:
why not?

Reply:
Oh, I thought the theming was obsolete.  We can use it if you want though.  Post-beta TODO: discuss further

Reply:
See https://github.com/Microsoft/pxt/issues/3119